### PR TITLE
DOCA Source Stage improvements

### PR DESCRIPTION
**Pull request by e-ago**
Upgrade DOCA source stage API from 2.2 to 2.5 + code restyle

```[tasklist]
### Waiting on
- [ ] https://github.com/nv-morpheus/MRC/pull/433
- [ ] https://github.com/nv-morpheus/Morpheus/pull/1468
```

### CODE REVIEW

File: morpheus/stages/doca/doca_source_stage.py [hunk 1]
```diff
@@ -67,6 +68,9 @@ def __init__(
         self._max_concurrent = c.num_threads
         self._nic_pci_address = nic_pci_address
         self._gpu_pci_address = gpu_pci_address
+        self._traffic_type = traffic_type
+        if self._traffic_type != 'udp' and self._traffic_type != 'tcp':
+            raise NotImplementedError("The Morpheus DOCA source stage allows a only udp or tcp types of traffic flow " + self._traffic_type)
```
Comment (cwharris) [inline morpheus/stages/doca/doca_source_stage.py:73]:
```suggestion
        self._traffic_type = traffic_type.lower()
        if self._traffic_type not in ('udp', 'tcp'):
            raise NotImplementedError("The Morpheus DOCA source stage allows a only udp or tcp types of traffic flow " + traffic_type)
```

File: morpheus/stages/doca/doca_source_stage.py [hunk 2]
```diff
@@ -86,6 +90,9 @@ def supports_cpp_node(self):
     def _build_source(self, builder: mrc.Builder) -> mrc.SegmentObject:
 
         if self._build_cpp_node():
-            return self._doca_source_class(builder, self.unique_name, self._nic_pci_address, self._gpu_pci_address)
+            # return self._doca_source_class(builder, self.unique_name, self._nic_pci_address, self._gpu_pci_address, self._traffic_type)
```
Comment (cwharris) [inline morpheus/stages/doca/doca_source_stage.py:93]: Remove comment.

File: morpheus/_lib/doca/src/doca_source.cpp [hunk 3]
```diff
@@ -68,158 +68,290 @@ std::optional<uint32_t> ip_to_int(std::string const& ip_address)
...
+        std::vector<rmm::device_uvector<uint16_t>> dst_port_out_d;
+        std::vector<rmm::device_uvector<int32_t>> tcp_flags_out_d;
+        std::vector<rmm::device_uvector<int32_t>> ether_type_out_d;
+        std::vector<rmm::device_uvector<int32_t>> next_proto_id_out_d;
+        std::vector<rmm::device_uvector<uint32_t>> timestamp_out_d;
+
+        int thread_idx = mrc::runnable::Context::get_runtime_context().rank();
+
+        if (thread_idx >= MAX_QUEUE) {
+            LOG(ERROR) << "Thread ID " << thread_idx << " bigger than MAX_QUEUE " << MAX_QUEUE << std::endl;
```
Comment (cwharris) [inline morpheus/_lib/doca/src/doca_source.cpp:128]: We should never get here unless we have a bug further up in the code, so let's do
```suggestion
            MORPHEUS_FAIL(MORPHEUS_CONCAT_STR("Thread ID " << thread_idx << " bigger than MAX_QUEUE " << MAX_QUEUE));
```

File: morpheus/_lib/doca/src/doca_source.cpp [hunk 4]
```diff
@@ -68,158 +68,290 @@ std::optional<uint32_t> ip_to_int(std::string const& ip_address)
...
-
-            cudaStreamSynchronize(rmm::cuda_stream_default);
-
-            auto packet_count = packet_count_d.value(rmm::cuda_stream_default);
-
-            if (packet_count == 0)
-            {
-                continue;
+            /*
+             * Can't run the received kernel in persistent mode for a weirdness in the rxcpp/mrc framework.
```
Comment (cwharris) [inline morpheus/_lib/doca/src/doca_source.cpp:214]: Please open an issue in the Morpheus github repository with this comment and any other pertinent information, and remove this comment from the code.

File: morpheus/_lib/doca/src/doca_source_kernels.cu [hunk 5]
```diff
@@ -157,247 +190,271 @@ __device__ int64_t mac_bytes_to_int64(uint8_t* mac)
...
+
+            // printf("CUDA rx time %ld proc time %ld pkt conv %ld block reduce %ld\n",
+            //         rx_stop - rx_start,
+            //         pkt_proc - rx_stop,
+            //         reduce_start - rx_stop,
+            //         reduce_stop - reduce_start);
+        }
+        __syncthreads();
+
+#if 0
```
Comment (cwharris) [inline morpheus/_lib/doca/src/doca_source_kernels.cu:422]: Please delete unused code.

File: morpheus/_lib/doca/src/doca_source_kernels.cu [hunk 6]
```diff
@@ -503,12 +560,13 @@ std::unique_ptr<cudf::column> integers_to_mac(
 
   auto const_17_itr = thrust::constant_iterator<cudf::size_type>(17);
   // this transform iterator is a workaround for a compilation issue caused by passing a plain constant iteartor
-  auto offsets_transformer = [] __device__(auto item) -> cudf::size_type { return item;};
-  auto offsets_transformer_itr = thrust::make_transform_iterator(const_17_itr, offsets_transformer);
+  // No error anymmore?
```
Comment (cwharris) [inline morpheus/_lib/doca/src/doca_source_kernels.cu:563]: Please delete commented code.

File: morpheus/_lib/doca/src/doca_source_kernels.cu [hunk 7]
```diff
@@ -626,12 +644,18 @@ std::unique_ptr<cudf::column> gather_payload(
...
+  // cuDeviceGet(&cuDevice, 0);
+  // cuCtxCreate(&cuContext, CU_CTX_SCHED_SPIN|CU_CTX_MAP_HOST, cuDevice);
+  // cuCtxPushCurrent(cuContext);
   auto [offsets_column, bytes] = cudf::detail::make_offsets_child_column(
     payload_sizes,
     payload_sizes + packet_count,
     stream,
     mr
   );
+  // cuCtxPopCurrent(&cuContext);
```
Comment (cwharris) [inline morpheus/_lib/doca/src/doca_source_kernels.cu:658]: Please delete unused code.

**Review by cwharris — CHANGES_REQUESTED** [review]
Changes look good outside of styles and leftover commented code. To get this passing CI you will need to run `./ci/scripts/fix_all.sh` from the Morpheus repository root with the Morpheus conda environment activated. That will adjust code styles, which means you will need to make a new commit. If there are still issues with CI, we can look in to them more closesly, and depending on how involved they are I may be able to make the updates.

File: morpheus/_lib/doca/src/doca_source.cpp (hunk 4 above)
Comment (e-ago) [reply morpheus/_lib/doca/src/doca_source.cpp:214]: https://github.com/nv-morpheus/Morpheus/issues/1557

File: morpheus/_lib/doca/src/doca_source_kernels.cu (hunk 5 above)
Comment (e-ago) [reply morpheus/_lib/doca/src/doca_source_kernels.cu:422]: added an if macro as this is the code to run it persistently (which can be useful to turn on for debug)

**Comment by e-ago** [comment]
Created 3 issues to keep track of the status: 2 bugs and 1 feature

https://github.com/nv-morpheus/Morpheus/issues/1557
https://github.com/nv-morpheus/Morpheus/issues/1558
https://github.com/nv-morpheus/Morpheus/issues/1559

Addressed all commits, can't run CI. Sent error stack via Slack message

**Review by cwharris — APPROVED** [review]

**Pull request merged**